smarts/core/sensors.py

```diff
@@ -1013,3 +1042,62 @@ def __call__(self, linear_velocity, angular_velocity):
 
     def teardown(self):
         pass
+
+
+class ViaSensor(Sensor):
+    def __init__(self, vehicle, mission_planner, lane_acquisition_range):
+        self._consumed_via_points = set()
+        self._mission_planner: MissionPlanner = mission_planner
+        self._acquisition_range = lane_acquisition_range
+        self._vehicle = vehicle
+
+    @property
+    def _vias(self) -> Iterable[ScenarioVia]:
+        return self._mission_planner.mission.via
+
+    def __call__(self):
+        near_points: List[ViaPointData] = list()
+        hit_points: List[ViaPointData] = list()
+        vehicle_position = self._vehicle.position[:2]
+
+        @lru_cache()
+        def closest_point_on_lane(position, lane_id):
+            return self._mission_planner.closest_point_on_lane(position, lane_id)
+
+        for via in self._vias:
+            closest_position_on_lane = closest_point_on_lane(
+                tuple(vehicle_position), via.lane_id
+            )
+            closest_position_on_lane = closest_position_on_lane[:2]
+
+            dist_from_lane_sq = squared_dist(vehicle_position, closest_position_on_lane)
+            if dist_from_lane_sq > self._acquisition_range ** 2:
+                continue
+
+            point = ViaPointData(
+                tuple(via.via_position),
+                lane_index=via.lane_index,
+                edge_id=via.edge_id,
+                required_speed=via.required_speed,
+            )
+
+            near_points.append(point)
+            dist_from_point_sq = squared_dist(vehicle_position, via.via_position)
+            if (
+                dist_from_point_sq <= via.hit_distance ** 2
+                and via not in self._consumed_via_points
+                and self._vehicle.speed > via.required_speed
```

Comment:
How about something like: `np.isclose(via.required_speed, self._vehicle.speed, atol=1.5)`, otherwise the agent could just floor it and exceed these limits.

Reply:
Certainly, I added an accuracy parameter to the sensor.